### PR TITLE
Add the LifetimeDependenceInsertion pass.

### PR DESCRIPTION
**Pull request by atrick**
Insert mark_dependence [nonescaping] markers at every lifetime introducer that produces a lifetime-dependent value.

This enables diagnostics for lifetime dependencies under: -enable-experimental-feature NonescapableTypes

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test windows platform

**Comment by atrick** [comment]
Windows is not recognizing a registered unit test. I'm completely baffled.

@nate-chandler do you see any issue based on the links I posted here?

Windows output:
`Found no test named argument_conventions!`

Yet,
The test is declared here:
https://github.com/apple/swift/pull/71266/files#diff-e1ab05cb08ed4ec1af00937198c9040f2ed62768cfcc704883ad0fb919c7b400R1159

The test is registered here:
https://github.com/apple/swift/pull/71266/files#diff-112d1e2604aea4ea95802997b1b3290da09292fa0af8d5554b53ab5b7428bf34R139

The test is requested here:
https://github.com/apple/swift/pull/71266/files#diff-01ebe418db86530eece78f7a4098ccef0697c4e2d9b6de9facd881156e13196aR19

File: test/SILOptimizer/opaque_values_unit.sil [hunk 1]
```diff
@@ -7,15 +7,15 @@ class C {}
 // Verify that SILFunction_getSelfArgumentIndex respects SILFunctionConventions
 // and returns right value even in the presence of indirect results.
 //
```
Comment (nate-chandler) [inline test/SILOptimizer/opaque_values_unit.sil:9]:
```
// REQUIRES: swift_in_compiler
```

**Comment by nate-chandler** [comment]
@atrick I believe the issue is that the SwiftCompilerSources optimizer code doesn't get used on Windows, so `// REQUIRES: swift_in_compiler` is required in the test file.

**Comment by atrick** [comment]
> @atrick I believe the issue is that the SwiftCompilerSources optimizer code doesn't get used on Windows, so `// REQUIRES: swift_in_compiler` is required in the test file.

I just realized that myself :) Sorry for bothering you!

**Comment by atrick** [comment]
@swift-ci smoke test

**Comment by nate-chandler** [comment]
Added a note for troubleshooting this kind of failure: https://github.com/apple/swift/pull/71558

**Pull request merged**